dotnet/src/SemanticKernel/Orchestration/Plan.cs

```diff
@@ -225,22 +312,193 @@ public async Task<Plan> InvokeNextStepAsync(SKContext context)
+                    this.State.Set(item.Value, result.Result.Trim());
+                }
+                else if (result.Variables.Get(item.Key, out var value))
+                {
+                    this.State.Set(item.Value, value);
+                }
+            }
+            // if (string.IsNullOrEmpty(sequentialPlan.OutputKey))
+            // {
+            //     _ = this.State.Update(result.Result.Trim());
+            // }
+            // else
+            // {
+            //     this.State.Set(sequentialPlan.OutputKey, result.Result.Trim());
+            // }
+
+            // _ = this.State.Update(result.Result.Trim());
+            // if (!string.IsNullOrEmpty(sequentialPlan.OutputKey))
+            // {
+            //     this.State.Set(sequentialPlan.OutputKey, result.Result.Trim());
+            // }
+
+            // if (!string.IsNullOrEmpty(sequentialPlan.ResultKey))
+            // {
+            //     _ = this.State.Get(SkillPlan.ResultKey, out var resultsSoFar);
+            //     this.State.Set(SkillPlan.ResultKey,
+            //         string.Join(Environment.NewLine + Environment.NewLine, resultsSoFar, result.Result.Trim()));
+            // }
+
+            #endregion Update State
+
             this.NextStepIndex++;
-            this.State.Update(context.Result.Trim());
         }
 
         return this;
     }
 
-    protected void SetFunction(ISKFunction function)
+    /// <summary>
+    /// Get JSON representation of the plan.
+    /// </summary>
+    public string ToJson()
+    {
+        return JsonSerializer.Serialize(this);
+    }
+
+    /// <summary>
+    /// To help with reading plans from <see cref="Orchestration.ContextVariables"/>.
+    /// </summary>
+    /// <param name="json">JSON string representation of aPlan</param>
+    /// <param name="context">The context to use for function registrations.</param>
+    /// <returns>An instance of a Plan object.</returns>
+    public static Plan FromJson(string json, SKContext context)
+    {
+        var plan = JsonSerializer.Deserialize<Plan>(json, new JsonSerializerOptions() { IncludeFields = true }) ?? new Plan(string.Empty);
+
+        plan = GetRegisteredFunctions(plan, context);
+
+        return plan;
+    }
+
+    private static Plan GetRegisteredFunctions(Plan plan, SKContext context)
+    {
+        if (plan.Steps.Count == 0)
+        {
+            if (context.IsFunctionRegistered(plan.SkillName, plan.Name, out var skillFunction))
+            {
+                Verify.NotNull(skillFunction, nameof(skillFunction));
+                plan.SetFunction(skillFunction);
+            }
+        }
+        else
+        {
+            foreach (var step in plan.Steps)
+            {
+                GetRegisteredFunctions(step, context);
+            }
+        }
+
+        return plan;
+    }
+
+    private ContextVariables GetNextStepVariables(ContextVariables variables, Plan step)
+    {
+        // If passing to a PLAN, do not use this.Description
```

Comment:
This should be good, let's clean up the code and comments to make it crystal clear.